extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/manager/impl/StructureManagerImpl.java

```diff
@@ -110,22 +129,21 @@ public boolean generateZoneGroupNames(ConnectionManager connectionManager) {
 
     @Override
     public String getZoneName(int zoneID) {
-        if (this.zoneGroupIdNameMap == null) {
+        if (zoneGroupIdNameMap == null) {
             return null;
         }
-        return this.zoneGroupIdNameMap.get(zoneID) != null ? (String) this.zoneGroupIdNameMap.get(zoneID)[0] : null;
+        return zoneGroupIdNameMap.get(zoneID) != null ? (String) zoneGroupIdNameMap.get(zoneID).zoneName : null;
     }
 
     @Override
-    @SuppressWarnings("unchecked")
     public String getZoneGroupName(int zoneID, short groupID) {
-        if (this.zoneGroupIdNameMap == null) {
+        if (zoneGroupIdNameMap == null) {
             return null;
         }
-        if (this.zoneGroupIdNameMap.get(zoneID) != null) {
-            if (this.zoneGroupIdNameMap.get(zoneID)[1] != null) {
-                if (((HashMap<Short, String>) this.zoneGroupIdNameMap.get(zoneID)[1]).get(groupID) != null) {
-                    return ((HashMap<Short, String>) this.zoneGroupIdNameMap.get(zoneID)[1]).get(groupID);
+        if (zoneGroupIdNameMap.get(zoneID) != null) {
+            if (zoneGroupIdNameMap.get(zoneID).groupIdNames != null) {
+                if (zoneGroupIdNameMap.get(zoneID).groupIdNames.get(groupID) != null) {
+                    return zoneGroupIdNameMap.get(zoneID).groupIdNames.get(groupID);
```

Comment:
Any reason that we call the `get` method multiple times?

`zoneGroupIdNameMap.get(zoneID)` is called four times (one time should be enough)
`groupIdNames.get(groupID)` is called twice (onen time should be enough)

So, instead of call get and if it is non null call it again, it would be better to store the result to a local variable and reuse it.

Reply:
Currently I think all if-blocks can be deleted or rather reduce to one. 
Because of the new object, the HashMap for the group-names can't be null any more.

So the code can look like that:
`return zoneGroupIdNameMap.get(zoneID) != null ? zoneGroupIdNameMap.get(zoneID).groupIdNames.get(groupID) : null;`

Or with a local variable for `zoneGroupIdNameMap.get(zoneID)`. But I’m not really sure which way is more efficient.
The local variable has to be created every time. On the other side, two HashMap accesses are only, if it is not null.

And lines [176](https://github.com/eclipse/smarthome/pull/3182/files/a6b88cd6d4fc8225bb2f6e828ce380cd4c7b1c1b#diff-15ed87cab5c37f4262bc4152fa0ea898R176) - [181](https://github.com/eclipse/smarthome/pull/3182/files/a6b88cd6d4fc8225bb2f6e828ce380cd4c7b1c1b#diff-15ed87cab5c37f4262bc4152fa0ea898R181) can be changed to:

`ZoneGroupsNameAndIDMap zoneGroupsNameID = zoneGroupNameIdMap.get(zoneName);`
`        if (zoneGroupsNameID != null) {`
`            return zoneGroupNameIdMap.get(zoneName).groupNameIds.get(groupName) != null
                    ? zoneGroupNameIdMap.get(zoneName).groupNameIds.get(groupName)
                    : -1;`

Reply:
Hm, I am not sure if I understand your doubts.

```Java
return map.get(key1) != null ? map.get(key1).get(key2) : null;
```

* if the first `map.get(key1)` is equal to `null` you have one `get`
* if the first `map.get(key1)` differs to `null` you have
** one additional `get` and a NPE if your key1 has been removed from the map by another thread (perhaps not possible in your impl. - I don't know). So 2x `get` + 1x NPE
** two additional `get`. So 3x `get`

```Java
final Map<...,...> tmp = map.get(key1);
return tmp != null ? tmp.get(key2) : null;
```

* one get if tmp equals to `null`
* one additional get if tmp differs to `null`
* you have one additional variable in the code, but this should be inlined by the byte compiler (AFAIK)

Reply:
I didn’t respect the case, that the first key can be removed by another thread. I think a NPE is improbable, but not impossible. So the local variable is safer and the better way. Thanks for the explanation.

In my opinion it is better when I add getter methods at the ZoneGroupsNameAndIDMap object to get the group-IDs and names. That do it more clearly that this maps can not be change.

I think it is useful when I incorporate the last comments etc. and commit it. That we get the same stand.